File: python_cpp/uPIMulator_frontend/src/abi/isa/cc/mul_nz_cc.py
from typing import Set

from abi.isa.cc._base_cc import BaseCC
from abi.isa.instruction.condition import Condition


class MulNZCC(BaseCC):
    def __init__(self, condition: Condition):
        super().__init__(
            MulNZCC.conditions(), condition,
        )

    @staticmethod
    def conditions() -> Set[Condition]:
        return {
            Condition.Z,
            Condition.NZ,
            Condition.XZ,
            Condition.XNZ,
            Condition.SZ,
            Condition.SNZ,
            Condition.SPL,
            Condition.SMI,
            Condition.LARGE,
            Condition.SMALL,
            Condition.TRUE,
        }

File: golang/uPIMulator/src/linker/kernel/instruction/cc/mul_nz_cc.go
package cc

import (
	"errors"
)

type MulNzCc struct {
	condition Condition
}

func (this *MulNzCc) Init(condition Condition) {
	conditions := map[Condition]bool{
		Z:     true,
		NZ:    true,
		XZ:    true,
		XNZ:   true,
		SZ:    true,
		SNZ:   true,
		SPL:   true,
		SMI:   true,
		LARGE: true,
		SMALL: true,
		TRUE:  true,
	}

	if _, found := conditions[condition]; !found {
		err := errors.New("condition is not allowed")
		panic(err)
	}

	this.condition = condition
}

func (this *MulNzCc) Condition() Condition {
	return this.condition
}
